### PR TITLE
Refactor component tasks

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Refactor component tasks
This is required for:
1. Being able to just use `StartTask` directly in NS without custom wrapping
2. Being able to properly test tasks (usage is coming in upcoming NS testing PR)

#### Changes

- Refactor component tasks and configuration

#### Testing

Integration testing

##### Regressions

Possibly incorrect jitter/backoff values used

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI according to the compatibility commitments in `README.md` for the chosen target branch.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

**Comment by rvolosatovs** [comment]
Blocking upcoming NS testing PR (#3052 )

**Review by johanstokking — COMMENTED** [review]
Code LGTM, but this is not a "quickfix" but needed a (technical debt) issue to close.

> Possibly incorrect jitter/backoff values used

Can you elaborate on this?

File: pkg/component/tasks.go [hunk 1]
```diff
@@ -37,85 +37,123 @@ const (
...
 
+func (c *Component) NewTaskConfig(id string, fn TaskFunc, restart TaskRestart, backoff *TaskBackoffConfig) *TaskConfig {
+	return &TaskConfig{
+		Context: c.Context(),
+		ID:      id,
+		Func:    fn,
+		Restart: restart,
+		Backoff: backoff,
+	}
+}
```
Comment (htdvisser) [inline pkg/component/tasks.go]: I don't think this adds much

**Review by htdvisser — APPROVED** [review]
LGTM

**Comment by rvolosatovs** [comment]
> Code LGTM, but this is not a "quickfix" but needed a (technical debt) issue to close.

This was necessary for NS testing, which is in turn necessary for registry optimizations, which is medium priority.

> > Possibly incorrect jitter/backoff values used
> 
> Can you elaborate on this?

Before packages used plain ("magic") constants when configuring tasks, this does not happen anymore - while refactoring I might have missed some value/set it to some incorrect value. It seemed like the default jitter value was used everywhere though, so doesn't seem likely to happen.

**Comment by johanstokking** [comment]
> > Code LGTM, but this is not a "quickfix" but needed a (technical debt) issue to close.
> 
> This was necessary for NS testing, which is in turn necessary for registry optimizations, which is medium priority.

How is this comment related to my comment? That because of medium priority there's no time to file a technical debt issue? Because that is not the way it works. Priority goes via priority labels only.

**Review by johanstokking — APPROVED** [review]

**Comment by rvolosatovs** [comment]
> > > Code LGTM, but this is not a "quickfix" but needed a (technical debt) issue to close.
> > 
> > 
> > This was necessary for NS testing, which is in turn necessary for registry optimizations, which is medium priority.
> 
> How is this comment related to my comment? That because of medium priority there's no time to file a technical debt issue? Because that is not the way it works. Priority goes via priority labels only.

I would need to stop working what I was working on(so make a huge context switch and contribute to making the main branch out-of-date - e.g. the event changes took some time to adapt to, the longer we wait, the more of such issues arise), file an issue (with "discussion" label I suppose?) and wait for a response. "just" filing an issue and immediately working on it is no different from what is done here.
While I agree that this should be have been filed in an issue, I don't see how this could realistically fit in the development process:
1. Refactor registry API as part of #2665 
2. Iteratively arrive at a viable implementation
3. Adapt tests - (until 1. and 2. are done, it makes no sense to do that, since final API is not known)
4. Find out that with existing implementation it is not possible to correctly test downlink due to testing changes (and see clear drawbacks of existing implementation, which e.g. make code more complex and disallow proper usage in NS)

--- Here the actual implementation of the feature is done and the tests are in-progress, component task API being the blocking issue
Your suggestion would be to file an issue about what I'm supposed to do (refactor component task API) and completely stop working on this feature waiting for feedback on the filed issue and find some, possibly merge-conflict-inducing, lower-priority issue to work on?

This does not seem like a viable choice to me - it directly hurts productivity.
Do you disagree?
Is there some other way how you would expect this to happen?

**Pull request merged**